src/System.Management.Automation/engine/parser/TypeInferenceVisitor.cs

```diff
@@ -987,154 +1024,179 @@ private IEnumerable<PSTypeName> InferTypesFrom(MemberExpressionAst memberExpress
+                    res.Add(type);
+                }
+            }
+            return res;
+        }
+
+        private List<PSTypeName> GetTypesOfMembers(string memberName, IList<object> members, ref bool maybeWantDefaultCtor, bool isInvokeMemberExpressionAst)
+        {
+            var memberNamesToCheck = new List<string> { memberName };
+            var res = new List<PSTypeName>(10);
+
+            AddTypesOfMembers(memberNamesToCheck, members, ref maybeWantDefaultCtor, isInvokeMemberExpressionAst, res);
+            return res;
+        }
+
+        private void AddTypesOfMembers(List<string> memberNamesToCheck, IList<object> members, ref bool maybeWantDefaultCtor, bool isInvokeMemberExpressionAst, List<PSTypeName> result)
+        {
+            for (int i = 0; i < memberNamesToCheck.Count; i++)
+            {
+                string memberNameToCheck = memberNamesToCheck[i];
+                foreach (var member in members)
+                {
+                    if (TryGetTypeFromMember(member, memberNameToCheck, ref maybeWantDefaultCtor, isInvokeMemberExpressionAst, result, memberNamesToCheck))
                     {
-                        var isInvokeMemberAst = memberExpressionAst is InvokeMemberExpressionAst;
-                        switch (member)
-                        {
-                            case PropertyInfo propertyInfo: // .net property
-                            {
-                                if (propertyInfo.Name.EqualsOrdinalIgnoreCase(memberName) && !isInvokeMemberAst)
-                                {
-                                    yield return new PSTypeName(propertyInfo.PropertyType);
-                                    goto NextMember;
-                                }
-                                continue;
-                            }
-                            case FieldInfo fieldInfo:       // .net field
-                            {
-                                if (fieldInfo.Name.EqualsOrdinalIgnoreCase(memberName) && !isInvokeMemberAst)
-                                {
-                                    yield return new PSTypeName(fieldInfo.FieldType);
-                                }
-                                continue;
-                            }
+                        goto NextMemberNameToCheck;
+                    }
+                }
+                NextMemberNameToCheck: { }
+            }
+        }
 
-                            case DotNetAdapter.MethodCacheEntry methodCacheEntry: // .net method
+        private bool TryGetTypeFromMember(object member, string memberName, ref bool maybeWantDefaultCtor, bool isInvokeMemberExpressionAst, List<PSTypeName> result, List<string> memberNamesToCheck)
+        {
+            switch (member)
+            {
+                case PropertyInfo propertyInfo: // .net property
+                {
+                    if (propertyInfo.Name.EqualsOrdinalIgnoreCase(memberName) && !isInvokeMemberExpressionAst)
```

Comment:
When casually reading code, I care more about the strings being compared than the comparer, I might skip over the comparer completely. Written this way, I stumble over the long name before getting to what I was looking for.

I suppose I could get used to this name, but only if it was used more widely.

---
In reply to: [122571659](https://github.com/PowerShell/PowerShell/pull/3972#discussion_r122571659) [](ancestors = 122571659)